# miRge-Py3

## Main goals
- [ ] Easy to implement and use
- [ ] Robust (accurate) data output
- [ ] Fast and able to run multiple datasets at once
- [ ] Attractive and meaningful output

## Basic functions - command line options
- [ ] User friendly parameters 
- [ ] Implement [`Read the docs`](https://readthedocs.org/)
- [ ] `Additional features here`

## Primary functions - Annotation:
- [ ] Incorporate latest version of cutadapt and its features. 
- [ ] Keep track to update features of miRge automatically with newer versions of cutadapt.
- [ ] No installation required for [Bowtie 1.2.3](https://sourceforge.net/projects/bowtie-bio/files/bowtie/1.2.3/). Use: `\bin\bowtie`
- [ ] `Additional features here`

## Secondary functions - Predict:
- [ ] RNA-fold for novel miRNAs
- [ ] Correct miRge GFF output
- [ ] tRNA halve and fragment prediction
- [ ] SVM re-implementation 
- [ ] PDF and HTML output
- [ ] Incorporate SQLite db for quick query and visualization
- [ ] `Add features here`

## Additional features - Customization of software:
- [ ] Bowtie to avoid mismatches in the seed region - Interact with **Mihaela Pertea**. 
- [ ] Software to use .SRA, .fastq.gz and .fastq
- [ ] `Additional features here`

## Advanced features - Packaging and testing
- [ ] Integrate Cython (better performance) 
- [ ] CPU and GPU switch usage options 
- [ ] Interactive or visually appealing outputs 
- [ ] Docker implementation 
- [ ] Integrates with other high througput Genomics tools, especially having output work in a Bioconductor environment.
- [ ] GUI - [Beeware](https://beeware.org/project/using/desktop-app/) or [Electron](https://electronjs.org/) implementaion which ever works best for cross platform
- [ ] GUI requires an installation of `Windows Subsystem for Linux`
  - [ ] [Cygwin](https://www.cygwin.com/) for Windows < 10
  - [ ] [Ubuntu](https://docs.microsoft.com/en-us/windows/wsl/install-win10) for Windows10
- [ ] `Additional features here`

#### References: 
- [miRge2.0](https://github.com/mhalushka/miRge)
- [miRge1.0](https://github.com/mhalushka/miRge-1) 
  - [sample_output](https://baraslab.github.io/miRge/miRge/miRge.exampleOutput/report.html)
- [sRNAtoolbox](https://bioinfo5.ugr.es/srnatoolbox/srnabench/)
- [Tools4miRs](https://tools4mirs.org/software/isomirs_identification/)
- [Chimira](http://wwwdev.ebi.ac.uk/enright-dev/chimira/index.php)
- [Beeware by Russell Keith-Magee](https://www.youtube.com/watch?v=qaPzlIJ57dk) 
- [ViennaRNA 2.0](https://github.com/ViennaRNA/ViennaRNA)
- [ ] `Additional features here`

## Upcoming events:
- [COVALENCE CONFERENCE 2020](https://www.covalenceconf.com/)
  - Date: January 24, 2020. (09:00AM to 07:00PM)
  - Venue:
    > Slack HQ, 
    > 500 Howard St,
    > San Francisco CA 94105.
