

<!DOCTYPE html>
<html class="writer-html5" lang="python" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>User guide &mdash; miRge3 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="miRge3.0 output" href="mirge_output.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> miRge3
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">User guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mirge3-0-libraries">miRge3.0 libraries</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#command-line-interface-cli">Command-line Interface (CLI)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#direct-download">Direct download</a></li>
<li class="toctree-l3"><a class="reference internal" href="#graphical-user-interface-gui">Graphical User Interface (GUI)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#building-new-libraries">Building new libraries</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cli-example-usage">CLI - Example usage</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#test">Test</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trimming-both-5-and-3-adapters-linked-adapters">Trimming both 5' and 3' adapters - Linked adapters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#running-samples-with-umi">Running samples with UMI</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#qiagen-based-umi">Qiagen - based UMI</a></li>
<li class="toctree-l4"><a class="reference internal" href="#illumina-based-umi-4n-method">Illumina - based UMI/4N method</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#performing-differential-expression-analysis">Performing differential expression analysis</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#mirge3-0-gui">miRge3.0 GUI</a></li>
<li class="toctree-l2"><a class="reference internal" href="#resources">Resources</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="mirge_output.html">miRge3.0 output</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">Frequently asked question (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">MIT License</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">miRge3</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>User guide</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/arunhpatil/miRge3.0/blob/master/docs/source/quick_start.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="user-guide">
<h1>User guide<a class="headerlink" href="#user-guide" title="Permalink to this headline">¶</a></h1>
<div class="section" id="parameters">
<h2>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline">¶</a></h2>
<p>To view command-line parameters type <code class="docutils literal notranslate"><span class="pre">miRge3.0</span> <span class="pre">-h</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>usage: miRge3.0 [options]

miRge3.0 (Comprehensive analysis of small RNA sequencing Data)

optional arguments:
  -h, --help  show this help message and exit
  --version   show program&#39;s version number and exit

Options:
  -s,    --samples            list of one or more samples separated by comma or a file with list of samples separated by new line (accepts *.fastq, *.fastq.gz) 
  -db,   --mir-DB             the reference database of miRNA. Options: miRBase and miRGeneDB (Default: miRBase) 
  -lib,  --libraries-path     the path to miRge libraries 
  -on,   --organism-name      the organism name can be human, mouse, fruitfly, nematode, rat or zebrafish
  -ex,   --crThreshold        the threshold of the proportion of canonical reads for the miRNAs to retain. Range for ex (0 - 0.5), (Default: 0.1)
  -phr,  --phred64            phred64 format (Default: 33)
  -spk,  --spikeIn            switch to annotate spike-ins if spike-in bowtie index files are located at the path of bowtie&#39;s index files (Default: off)
  -ie,   --isoform-entropy    switch to calculate isomir entropy (default: off)
  -cpu,  --threads            the number of processors to use for trimming, qc, and alignment (Default: 1)
  -ai,   --AtoI               switch to calculate A to I editing (Default: off)
  -tcf   --tcf-out            switch to write trimmed and collapsed fasta file (Default: off)
  -gff   --gff-out            switch to output isomiR results in gff format (Default: off) 
  -bam   --bam-out            switch to output results in bam format (Default: off) 
  -trf   --tRNA-frag          switch to analyze tRNA fragment and halves (Default: off)
  -o     --outDir             the directory of the outputs (Default: current directory) 
  -dex   --diffex             perform differential expression with DESeq2 (Default: off)
  -mdt   --metadata           the path to metadata file (Default: off, require &#39;.csv&#39; file format if -dex is opted)
  -cms   --chunkmbs           chunk memory in megabytes per thread to use during bowtie alignment (Default: 256)
  -shh   --quiet              enable quiet/silent mode, only show warnings and errors (Default: off)

Data pre-processing:
  -a,    --adapter            Sequence of a 3&#39; adapter. The adapter and subsequent bases are trimmed
  -g,    --front              Sequence of a 5&#39; adapter. The adapter and any preceding bases are trimmed
  -u,    --cut                Remove bases from each read. If LENGTH is positive, remove bases from the beginning. If LENGTH is negative, remove bases from the end
  -nxt,  --nextseq-trim       NextSeq-specific quality trimming (each read). Trims also dark cycles appearing as high-quality G bases
  -q,    --quality-cutoff     Trim low-quality bases from 5&#39; and/or 3&#39; ends of each read before adapter removal. If one value is given, only the 3&#39; end is trimmed
                              If two comma-separated cutoffs are given, the 5&#39; end is trimmed with the first cutoff, the 3&#39; end with the second
  -l,    --length             Shorten reads to LENGTH. Positive values remove bases at the end while negative ones remove bases at the beginning. This and the following
                              modifications are applied after adapter trimming
  -NX,   --trim-n             Trim N&#39;s on ends of reads
  -m,    --minimum-length     Discard reads shorter than LEN. (Default: 16)
  -umi,  --uniq-mol-ids       Trim nucleotides of specific length at 5’ and 3’ ends of the read, after adapter trimming. eg: 4,4 or 0,4. (Use -udd to remove PCR duplicates)  
  -udd,  --umiDedup           Specifies argument to removes PCR duplicates (Default: False); if TRUE it will remove UMI and remove PCR duplicates otherwise it only remove UMI and keep the raw counts (Require -umi option)
  -qumi, --qiagenumi          Removes PCR duplicates of reads obtained from Qiagen platform (Default: Illumina; &quot;-umi x,y &quot; Required)
  
  
miRNA Error Correction:
  microRNA correction method for single base substitutions due to sequencing errors (Note: Refines reads at the expense of time)
  -mEC,  --miREC              Enable miRNA error correction (miREC)
  -kh,   --threshold          the value for frequency threshold τ (Default kh = 5)
  -ks,   --kmer-start         kmer range start value (k_1, default 15) 
  -ke,   --kmer-end           kmer range end value (k_end, default 20)

Predicting novel miRNAs:
  The predictive model for novel miRNA detection is trained on human and mouse!
  -nmir, --novel-miRNA        include prediction of novel miRNAs
  -minl, --minLength          the minimum length of the retained reads for novel miRNA detection (default: 16)
  -maxl, --maxLength          the maximum length of the retained reads for novel miRNA detection (default: 25)
  -c,    --minReadCounts      the minimum read counts supporting novel miRNA detection (default: 2)
  -mloc, --maxMappingLoci     the maximum number of mapping loci for the retained reads for novel miRNA detection (default: 3)
  -sl,   --seedLength         the seed length when invoking Bowtie for novel miRNA detection (default: 25)
  -olc,  --overlapLenCutoff   the length of overlapped seqence when joining reads into longer sequences based on the coordinate 
                              on the genome for novel miRNA detection (default: 14)
  -clc,  --clusterLength      the maximum length of the clustered sequences for novel miRNA detection (default: 30)

Optional PATH arguments:
  -pbwt, --bowtie-path        the path to system&#39;s directory containing bowtie binary
  -psam, --samtools-path      the path to system&#39;s directory containing samtools binary
  -prf,  --RNAfold-path       the path to system&#39;s directory containing RNAfold binary
</pre></div>
</div>
</div>
<div class="section" id="mirge3-0-libraries">
<h2>miRge3.0 libraries<a class="headerlink" href="#mirge3-0-libraries" title="Permalink to this headline">¶</a></h2>
<p>miRge3.0 pipeline aligns the raw reads against a set of small-RNA annotation libraries. The libraries specific to the organism of interest can be obtained from <a class="reference external" href="https://sourceforge.net/projects/mirge3/files/miRge3_Lib/">SourceForge</a>. Downloading the libraries on terminal:</p>
<div class="section" id="command-line-interface-cli">
<h3>Command-line Interface (CLI)<a class="headerlink" href="#command-line-interface-cli" title="Permalink to this headline">¶</a></h3>
<p>We recommend to create a directory <code class="docutils literal notranslate"><span class="pre">miRge3_Lib</span></code> and download using wget as shown below,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">miRge3_Lib</span>
<span class="n">cd</span> <span class="n">miRge3_Lib</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">human</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/miRge3_Lib/human.tar.gz/download&quot;</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">mouse</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/miRge3_Lib/mouse.tar.gz/download&quot;</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">rat</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/miRge3_Lib/rat.tar.gz/download&quot;</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">nematode</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/miRge3_Lib/nematode.tar.gz/download&quot;</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">fruitfly</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/miRge3_Lib/fruitfly.tar.gz/download&quot;</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">zebrafish</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/miRge3_Lib/zebrafish.tar.gz/download&quot;</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">hamster</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/miRge3_Lib/hamster.tar.gz/download&quot;</span>
</pre></div>
</div>
<p>Users can download only what is necessary. Unzip the files once downloaded by the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tar</span> <span class="o">-</span><span class="n">xzf</span> <span class="n">human</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Replace <code class="docutils literal notranslate"><span class="pre">human</span></code> with the organism of interest. If you want to extract all the files at once, you could use <code class="docutils literal notranslate"><span class="pre">tar</span> <span class="pre">-xzf</span> <span class="pre">*.tar.gz</span></code> instead.</p>
</div>
<div class="section" id="direct-download">
<h3>Direct download<a class="headerlink" href="#direct-download" title="Permalink to this headline">¶</a></h3>
<p>If you are having trouble downloading files through SourceForge, please use the direct link to download the library by clicking on links:
<a class="reference external" href="https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/mhalush1_jh_edu/EYMFZao-I4pGrvXbooaW25UBfNaVyO60tn3B0NUPh1dDMA?e=he2DHq">Human</a>, <a class="reference external" href="https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/mhalush1_jh_edu/Ec0bxWFCC9RBgftejLXk9jkBN0iLhuOdiSoEW05ZzlZmFg?e=cu3nHd">Mouse</a>, <a class="reference external" href="https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/mhalush1_jh_edu/EVS8LiTpr2tKnMyVPPvEzJABkAVRoH2LTSdtYNyl15oOqA?e=7uuRI3">Rat</a>, <a class="reference external" href="https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/mhalush1_jh_edu/EXr64OP2uVNKo0Xf8J9wW9UBhc-nm4RZJYgqt5xSC9VOgQ?e=r19IL0">Zebrafish</a>, <a class="reference external" href="https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/mhalush1_jh_edu/EUdmz6-PGb9Omy8rJt_zemUBqeIj_T_qGTPKE5wI8o6NRw?e=ets81v">Nematode</a>, <a class="reference external" href="https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/mhalush1_jh_edu/EfCYTKoxYfpKoUEl2iHC8FwBE8b3IOS4ArjMKQhUWH5Ujw?e=xhJxQH">Fruitfly</a>, <a class="reference external" href="https://livejohnshopkins-my.sharepoint.com/:u:/g/personal/mhalush1_jh_edu/EebwPHactwVOvw_g5aF1Dq0Bs6fmCNY20fuwSdPRz8IffQ?e=s8dm57">Golden Hamster</a> and <a class="reference external" href="https://livejohnshopkins-my.sharepoint.com/:t:/g/personal/mhalush1_jh_edu/EW6q7zO0GcJPjGvcpIoxRjEBIYBYdNv9d1oF3529v-Pwyg?e=EEer7g">md5sum</a>.</p>
</div>
<div class="section" id="graphical-user-interface-gui">
<h3>Graphical User Interface (GUI)<a class="headerlink" href="#graphical-user-interface-gui" title="Permalink to this headline">¶</a></h3>
<p>We recommend to create a folder <code class="docutils literal notranslate"><span class="pre">miRge3_Lib</span></code> and download the libraries directly from <a class="reference external" href="https://sourceforge.net/projects/mirge3/files/miRge3_Lib/">SourceForge</a>. Once downloaded, extract/unzip the compressed files.</p>
</div>
<div class="section" id="building-new-libraries">
<h3>Building new libraries<a class="headerlink" href="#building-new-libraries" title="Permalink to this headline">¶</a></h3>
<p>If you are interested in creating specific library for an organism that is not part of this set then please refer to <a class="reference external" href="https://github.com/mhalushka/miRge3_build">miRge3_build</a>.</p>
</div>
</div>
<div class="section" id="cli-example-usage">
<h2>CLI - Example usage<a class="headerlink" href="#cli-example-usage" title="Permalink to this headline">¶</a></h2>
<p>Example command usage:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">miRge3</span><span class="o">.</span><span class="mi">0</span> <span class="o">-</span><span class="n">s</span> <span class="n">SRR772403</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR772404</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR772405</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR772406</span><span class="o">.</span><span class="n">fastq</span> <span class="o">-</span><span class="n">lib</span> <span class="n">miRge3_Lib</span> <span class="o">-</span><span class="n">on</span> <span class="n">human</span> <span class="o">-</span><span class="n">db</span> <span class="n">mirgenedb</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_dir</span> <span class="o">-</span><span class="n">gff</span> <span class="o">-</span><span class="n">nmir</span> <span class="o">-</span><span class="n">trf</span> <span class="o">-</span><span class="n">ai</span> <span class="o">-</span><span class="n">cpu</span> <span class="mi">12</span> <span class="o">-</span><span class="n">a</span> <span class="n">illumina</span> 
</pre></div>
</div>
<p>Output command line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bowtie</span> <span class="n">version</span><span class="p">:</span> <span class="mf">1.2</span><span class="o">.</span><span class="mi">3</span>
<span class="n">Samtools</span> <span class="n">version</span><span class="p">:</span> <span class="mf">1.7</span>
<span class="n">RNAfold</span> <span class="n">version</span><span class="p">:</span> <span class="mf">2.4</span><span class="o">.</span><span class="mi">14</span>
<span class="n">Collecting</span> <span class="ow">and</span> <span class="n">validating</span> <span class="nb">input</span> <span class="n">files</span><span class="o">...</span>

<span class="n">miRge3</span><span class="o">.</span><span class="mi">0</span> <span class="n">will</span> <span class="n">process</span> <span class="mi">4</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">4</span> <span class="nb">input</span> <span class="n">file</span><span class="p">(</span><span class="n">s</span><span class="p">)</span><span class="o">.</span>

<span class="n">Cutadapt</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772403</span> <span class="ow">in</span> <span class="mf">2.5358</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Collapsing</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772403</span> <span class="ow">in</span> <span class="mf">0.0126</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Cutadapt</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772404</span> <span class="ow">in</span> <span class="mf">7.3542</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Collapsing</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772404</span> <span class="ow">in</span> <span class="mf">0.2786</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Cutadapt</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772405</span> <span class="ow">in</span> <span class="mf">11.0667</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Collapsing</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772405</span> <span class="ow">in</span> <span class="mf">0.8585</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Cutadapt</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772406</span> <span class="ow">in</span> <span class="mf">3.5771</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Collapsing</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772406</span> <span class="ow">in</span> <span class="mf">0.8677</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Matrix</span> <span class="n">creation</span> <span class="n">finished</span> <span class="ow">in</span> <span class="mf">0.3838</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Data</span> <span class="n">pre</span><span class="o">-</span><span class="n">processing</span> <span class="n">completed</span> <span class="ow">in</span> <span class="mf">27.2443</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Alignment</span> <span class="ow">in</span> <span class="n">progress</span> <span class="o">...</span>
<span class="n">Alignment</span> <span class="n">completed</span> <span class="ow">in</span> <span class="mf">15.8305</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Summarizing</span> <span class="ow">and</span> <span class="n">tabulating</span> <span class="n">results</span><span class="o">...</span>
<span class="n">The</span> <span class="n">number</span> <span class="n">of</span> <span class="n">A</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">I</span> <span class="n">editing</span> <span class="n">sites</span> <span class="k">for</span> <span class="ow">is</span> <span class="n">less</span> <span class="n">than</span> <span class="mi">10</span> <span class="n">so</span> <span class="n">that</span> <span class="n">no</span> <span class="n">heatmap</span> <span class="ow">is</span> <span class="n">drawn</span><span class="o">.</span>
<span class="n">Summary</span> <span class="n">completed</span> <span class="ow">in</span> <span class="mf">71.4691</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Predicting</span> <span class="n">novel</span> <span class="n">miRNAs</span>


<span class="n">Performing</span> <span class="n">prediction</span> <span class="n">of</span> <span class="n">novel</span> <span class="n">miRNAs</span><span class="o">...</span>
<span class="n">Start</span> <span class="n">to</span> <span class="n">predict</span>
<span class="n">Prediction</span> <span class="n">of</span> <span class="n">novel</span> <span class="n">miRNAs</span> <span class="n">Completed</span> <span class="p">(</span><span class="mf">104.83</span> <span class="n">sec</span><span class="p">)</span>

<span class="n">The</span> <span class="n">analysis</span> <span class="n">completed</span> <span class="ow">in</span> <span class="mf">222.2487</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="test">
<h3>Test<a class="headerlink" href="#test" title="Permalink to this headline">¶</a></h3>
<p>The test case illustrates the usage of miRge3.0 with a sample dataset, mapping to human reference libraries.</p>
<ul class="simple">
<li><p>First download human miRge libraries as shown below:</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">miRge3_Lib</span>
<span class="n">cd</span> <span class="n">miRge3_Lib</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">human</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/miRge3_Lib/human.tar.gz/download&quot;</span>
<span class="n">tar</span> <span class="o">-</span><span class="n">xzf</span> <span class="n">human</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
<span class="n">cd</span> <span class="o">..</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Download the sample file from Source Forge, <a class="reference external" href="https://sourceforge.net/projects/mirge3/files/test/SRR772403.fastq.gz/download">SRR772403</a></p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">You</span> <span class="n">can</span> <span class="n">download</span> <span class="n">to</span> <span class="n">your</span> <span class="n">working</span> <span class="n">directory</span> <span class="k">as</span> <span class="n">shown</span> <span class="n">below</span><span class="p">:</span>
<span class="n">wget</span> <span class="o">-</span><span class="n">O</span> <span class="n">SRR772403</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="s2">&quot;https://sourceforge.net/projects/mirge3/files/test/SRR772403.fastq.gz/download&quot;</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Run basic miRge3.0 command to annotate and report isomiRs</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">miRge3</span><span class="o">.</span><span class="mi">0</span> <span class="o">-</span><span class="n">s</span> <span class="n">SRR772403</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">lib</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">d</span><span class="o">/</span><span class="n">Halushka_lab</span><span class="o">/</span><span class="n">Arun</span><span class="o">/</span><span class="n">miRge3_Lib</span> <span class="o">-</span><span class="n">a</span> <span class="n">illumina</span> <span class="o">-</span><span class="n">on</span> <span class="n">human</span> <span class="o">-</span><span class="n">db</span> <span class="n">mirbase</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_dir</span> <span class="o">-</span><span class="n">gff</span> <span class="o">-</span><span class="n">cpu</span> <span class="mi">8</span>

<span class="n">bowtie</span> <span class="n">version</span><span class="p">:</span> <span class="mf">1.3</span><span class="o">.</span><span class="mi">0</span>
<span class="n">cutadapt</span> <span class="n">version</span><span class="p">:</span> <span class="mf">3.1</span>
<span class="n">Samtools</span> <span class="n">version</span><span class="p">:</span> <span class="mf">1.11</span>
<span class="n">Collecting</span> <span class="ow">and</span> <span class="n">validating</span> <span class="nb">input</span> <span class="n">files</span><span class="o">...</span>

<span class="n">miRge3</span><span class="o">.</span><span class="mi">0</span> <span class="n">will</span> <span class="n">process</span> <span class="mi">1</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">1</span> <span class="nb">input</span> <span class="n">file</span><span class="p">(</span><span class="n">s</span><span class="p">)</span><span class="o">.</span>

<span class="n">Cutadapt</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772403</span> <span class="ow">in</span> <span class="mf">3.4343</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="n">Collapsing</span> <span class="n">finished</span> <span class="k">for</span> <span class="n">file</span> <span class="n">SRR772403</span> <span class="ow">in</span> <span class="mf">0.0216</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Matrix</span> <span class="n">creation</span> <span class="n">finished</span> <span class="ow">in</span> <span class="mf">0.0263</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Data</span> <span class="n">pre</span><span class="o">-</span><span class="n">processing</span> <span class="n">completed</span> <span class="ow">in</span> <span class="mf">3.5111</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Alignment</span> <span class="ow">in</span> <span class="n">progress</span> <span class="o">...</span>
<span class="n">Alignment</span> <span class="n">completed</span> <span class="ow">in</span> <span class="mf">8.1488</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">Summarizing</span> <span class="ow">and</span> <span class="n">tabulating</span> <span class="n">results</span><span class="o">...</span>
<span class="n">Summary</span> <span class="n">completed</span> <span class="ow">in</span> <span class="mf">2.27</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>


<span class="n">The</span> <span class="n">analysis</span> <span class="n">completed</span> <span class="ow">in</span> <span class="mf">15.2276</span> <span class="n">second</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Output folder, sample output can be accessed <a class="reference external" href="https://sourceforge.net/projects/mirge3/files/test/output_dir/">here</a></p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">miRge</span> <span class="n">creates</span> <span class="n">a</span> <span class="n">subfolder</span> <span class="n">inside</span> <span class="n">the</span> <span class="n">folder</span> <span class="s2">&quot;output_dir&quot;</span> <span class="ow">and</span> <span class="nb">all</span> <span class="n">the</span> <span class="n">files</span> <span class="n">will</span> <span class="n">be</span> <span class="n">stored</span> <span class="n">there</span><span class="o">.</span> <span class="n">The</span> <span class="n">test</span> <span class="n">output</span> <span class="n">can</span> <span class="n">be</span> <span class="n">accessed</span> <span class="n">at</span> <span class="n">the</span> <span class="n">following</span> <span class="n">link</span><span class="p">:</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">sourceforge</span><span class="o">.</span><span class="n">net</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">mirge3</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">test</span><span class="o">/</span><span class="n">output_dir</span><span class="o">/</span><span class="n">miRge</span><span class="o">.</span><span class="mi">2021</span><span class="o">-</span><span class="mi">06</span><span class="o">-</span><span class="mi">25_15</span><span class="o">-</span><span class="mi">16</span><span class="o">-</span><span class="mi">58</span><span class="o">/</span>
</pre></div>
</div>
</div>
<div class="section" id="trimming-both-5-and-3-adapters-linked-adapters">
<h3>Trimming both 5' and 3' adapters - Linked adapters<a class="headerlink" href="#trimming-both-5-and-3-adapters-linked-adapters" title="Permalink to this headline">¶</a></h3>
<p>If the data contains adapters at both 5' and 3' ends of the reads and both the adapters need to be removed then you should perform linked adapter trimming. This is part of Cutadapt and more about linked adapters can be found <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/guide.html#linked-adapters-combined-5-and-3-adapter">here</a>.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">miRge3</span><span class="o">.</span><span class="mi">0</span> <span class="o">-</span><span class="n">s</span> <span class="n">DRR013811</span><span class="o">.</span><span class="n">fastq</span> <span class="o">-</span><span class="n">lib</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">d</span><span class="o">/</span><span class="n">Halushka_lab</span><span class="o">/</span><span class="n">Arun</span><span class="o">/</span><span class="n">GTF_Repeats_miRge2to3</span><span class="o">/</span><span class="n">miRge3_Lib</span><span class="o">/</span><span class="n">revised_hsa</span>  <span class="o">-</span><span class="n">on</span> <span class="n">human</span> <span class="o">-</span><span class="n">db</span> <span class="n">mirbase</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_dir</span> <span class="o">-</span><span class="n">g</span> <span class="s2">&quot;TTAGGC...TGGAATTCTCGGGTGCCAAGGAACTCCAGT&quot;</span>
</pre></div>
</div>
<p>Description of adapter: <code class="docutils literal notranslate"><span class="pre">&quot;TTAGGC...TGGAATTCTCGGGTGCCAAGGAACTCCAGT&quot;</span></code>, where <code class="docutils literal notranslate"><span class="pre">TTAGGC</span></code> is the 5' adapter and <code class="docutils literal notranslate"><span class="pre">TGGAATTCTCGGGTGCCAAGGAACTCCAGT</span></code> is the 3' adapter sequence.</p>
<p>Note: Complete adapter sequence must be provided (mandatory) i.e., simply specifying illumina will not be decoded to its actual adapter sequence. <br>
This will <code class="docutils literal notranslate"><span class="pre">NOT</span> <span class="pre">WORK</span></code>: -g <code class="docutils literal notranslate"><span class="pre">&quot;TTAGGC...illumina&quot;</span></code><br>
This will <code class="docutils literal notranslate"><span class="pre">WORK</span></code>: -g <code class="docutils literal notranslate"><span class="pre">&quot;TTAGGC...TGGAATTCTCGGGTGCCAAGGAACTCCAGT&quot;</span></code></p>
</div>
<div class="section" id="running-samples-with-umi">
<h3>Running samples with UMI<a class="headerlink" href="#running-samples-with-umi" title="Permalink to this headline">¶</a></h3>
<div class="section" id="qiagen-based-umi">
<h4>Qiagen - based UMI<a class="headerlink" href="#qiagen-based-umi" title="Permalink to this headline">¶</a></h4>
<p>Testing sample data run on UMI obtained from Qiagen platform. Important parameters are (<code class="docutils literal notranslate"><span class="pre">-umi</span></code>, <code class="docutils literal notranslate"><span class="pre">--qiagenumi</span></code> and <code class="docutils literal notranslate"><span class="pre">-udd</span></code>)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">miRge3</span><span class="o">.</span><span class="mi">0</span> <span class="o">-</span><span class="n">s</span> <span class="n">SRR13077007</span><span class="o">.</span><span class="n">fastq</span> <span class="o">-</span><span class="n">db</span> <span class="n">miRBase</span> <span class="o">-</span><span class="n">lib</span> <span class="n">miRge3_Lib</span> <span class="o">-</span><span class="n">on</span> <span class="n">human</span> <span class="o">-</span><span class="n">a</span> <span class="n">AACTGTAGGCACCATCAAT</span> <span class="o">--</span><span class="n">qiagenumi</span> <span class="o">-</span><span class="n">umi</span> <span class="mi">0</span><span class="p">,</span><span class="mi">12</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_dir</span> <span class="o">-</span><span class="n">cpu</span> <span class="mi">10</span> <span class="o">-</span><span class="n">udd</span>
</pre></div>
</div>
<p>Please note: As of July, 2021, the standard internal 3' adapter was AACTGTAGGCACCATCAAT ligated to 12 nucleotide UMI sequence followed by external 3' adapter sequence. If you have different internal adapter other than AACTGTAGGCACCATCAAT, then please provide that.</p>
<p>Example of reads, UMI and adapters for hsa-let-7a (sequence left to right in the order mentioned below with-in angular brackets):</p>
<p>&lt;hsa-let-7a-5p: TGAGGTAGTAGGTTGTATAGTT&gt;&lt;Internal 3' adapter:<code class="docutils literal notranslate"><span class="pre">AACTGTAGGCACCATCAAT</span></code>&gt;&lt;12 nt UMI&gt;&lt;external 3' adapter <code class="docutils literal notranslate"><span class="pre">AGATCGGAAGAGCACACGTCT</span></code>&gt;</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">TGAGGTAGTAGGTTGTATAGTTAACTGTAGGCACCATCAATGTTAGACCTGCAAGATCGGAAGAGCACACGTCTG</span>
<span class="n">TGAGGTAGTAGGTTGTATAGTTAACTGTAGGCACCATCAATCAATGACGATTTAGATCGGAAGAGCACACGTCTG</span>
<span class="n">TGAGGTAGTAGGTTGTATAGTTAACTGTAGGCACCATCAATAAACAAAGATCCAGATCGGAAGAGCACACGTCTG</span>
<span class="n">TGAGGTAGTAGGTTGTATAGTTAACTGTAGGCACCATCAATCGCATCGCCGACAGATCGGAAGAGCACACGTCTG</span>
<span class="n">TGAGGTAGTAGGTTGTATAGTTAACTGTAGGCACCATCAATTTTGCCATTACTAGATCGGAAGAGCACACGTCTG</span>
</pre></div>
</div>
</div>
<div class="section" id="illumina-based-umi-4n-method">
<h4>Illumina - based UMI/4N method<a class="headerlink" href="#illumina-based-umi-4n-method" title="Permalink to this headline">¶</a></h4>
<p>Testing sample data run on UMI/4N obtained from Illumina or similar platform. Important parameters are (<code class="docutils literal notranslate"><span class="pre">-umi</span></code> and <code class="docutils literal notranslate"><span class="pre">-udd</span></code>)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">miRge3</span><span class="o">.</span><span class="mi">0</span> <span class="o">-</span><span class="n">s</span> <span class="n">SRR6379839</span><span class="o">.</span><span class="n">fastq</span> <span class="o">-</span><span class="n">db</span> <span class="n">miRBase</span> <span class="o">-</span><span class="n">lib</span> <span class="n">miRge3_Lib</span> <span class="o">-</span><span class="n">on</span> <span class="n">human</span> <span class="o">-</span><span class="n">a</span> <span class="n">illumina</span> <span class="o">-</span><span class="n">umi</span> <span class="mi">4</span><span class="p">,</span><span class="mi">4</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_dir</span> <span class="o">-</span><span class="n">cpu</span> <span class="mi">10</span> <span class="o">-</span><span class="n">udd</span>
</pre></div>
</div>
<p>&lt;04 nt UMI&gt;&lt;hsa-let-7a-5p: TGAGGTAGTAGGTTGTATAGTT&gt;&lt;04 nt UMI&gt;&lt;3' adapter:<code class="docutils literal notranslate"><span class="pre">TGGAATTCTCGGGTGCCAAGGAACTCCAGTCACCGGAATATCTCG</span></code>&gt;</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">TACATGAGGTAGTAGGTTGTATAGTTCCTCTGGAATTCTCGGGTGCCAAGGAACTCCAGTCACCGGAATATCTCG</span>
<span class="n">TACCTGAGGTAGTAGGTTGTATAGTTACTATGGAATTCTCGGGTGCCAAGGAACTCCAGTCACCGGAATATCTCG</span>
<span class="n">CAGGTGAGGTAGTAGGTTGTATAGTTGGTATGGAATTCTCGGGTGCCAAGGAACTCCAGTCACCGGAATATCTCG</span>
<span class="n">AGAATGAGGTAGTAGGTTGTATAGTTACTATGGAATTCTCGGGTGACAAGGAACTCCAGTCACCGGAATATCTCG</span>
<span class="n">AGGTTGAGGTAGTAGGTTGTATAGTTACTATGGAATTCTCGGGTGCCAAGGAACTCCAGTCACCGGAATATCTCG</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="performing-differential-expression-analysis">
<h3>Performing differential expression analysis<a class="headerlink" href="#performing-differential-expression-analysis" title="Permalink to this headline">¶</a></h3>
<ol class="simple">
<li><p>Download example datasets from NCBI <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP182725">SRA</a> (Note: Tutorial on how to download SRA files is below).</p></li>
<li><p>Prepare metadata information in CSV format as shown below. For this tutorial, download the file from <a class="reference external" href="https://sourceforge.net/projects/mirge3/files/test/DESmetadata.csv/download">here</a>.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">id</span><span class="p">,</span><span class="n">group</span>
<span class="n">SRR8497647</span><span class="p">,</span><span class="n">Control</span>
<span class="n">SRR8497648</span><span class="p">,</span><span class="n">Control</span>
<span class="n">SRR8497649</span><span class="p">,</span><span class="n">Control</span>
<span class="n">SRR8497650</span><span class="p">,</span><span class="n">Control</span>
<span class="n">SRR8497651</span><span class="p">,</span><span class="n">treated</span>
<span class="n">SRR8497652</span><span class="p">,</span><span class="n">treated</span>
<span class="n">SRR8497653</span><span class="p">,</span><span class="n">treated</span>
<span class="n">SRR8497654</span><span class="p">,</span><span class="n">treated</span>
</pre></div>
</div>
<ol class="simple">
<li><p>Execute the following command:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">miRge3</span><span class="o">.</span><span class="mi">0</span> <span class="o">-</span><span class="n">s</span> <span class="n">SRR8497647</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR8497648</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR8497649</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR8497650</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR8497651</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR8497652</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR8497653</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span><span class="n">SRR8497654</span><span class="o">.</span><span class="n">fastq</span> <span class="o">-</span><span class="n">lib</span> <span class="n">miRge3_Lib</span> <span class="o">-</span><span class="n">on</span> <span class="n">human</span> <span class="o">-</span><span class="n">db</span> <span class="n">miRGeneDB</span> <span class="o">-</span><span class="n">o</span> <span class="n">differential_Exp</span> <span class="o">-</span><span class="n">a</span> <span class="n">TGGAATTCTCGG</span> <span class="o">-</span><span class="n">cpu</span> <span class="mi">12</span> <span class="o">-</span><span class="n">dex</span> <span class="o">-</span><span class="n">mdt</span> <span class="n">DESmetadata</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p>Tutorial on how to download SRA files:<br/>
This turorial is only brief introduction and doesn't cover all the details of downloading NCBI SRA files. You could find <a class="reference external" href="https://www.youtube.com/results?search_query=download+SRA+files">YouTube tutorials</a> on how to download SRA files.</p>
<ol class="simple">
<li><p>Download and install NCBI SRA toolkit: You could refer to NCBI <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK242621/">SRA Handbook</a> or <a class="reference external" href="https://github.com/ncbi/sra-tools">GitHub</a></p></li>
<li><p>Download command:
One could use <code class="docutils literal notranslate"><span class="pre">fasterq-dump</span> <span class="pre">-t</span> <span class="pre">temp</span> <span class="pre">-e</span> <span class="pre">10</span> <span class="pre">SRR8497647</span></code> or simply <code class="docutils literal notranslate"><span class="pre">fastq-dump</span> <span class="pre">SRR8497647</span></code>. The only difference being that the fasterq-dump is faster.
Similarly, download all other Runs (i.e., SRR8497648, SRR8497649 etc.)</p></li>
</ol>
</div>
</div>
<div class="section" id="mirge3-0-gui">
<h2>miRge3.0 GUI<a class="headerlink" href="#mirge3-0-gui" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>The application is cross platform, the image below is a screenshot of the software from MacOS
<img alt="_images/mac_exe.png" src="_images/mac_exe.png" /><br/><br/></p></li>
<li><p>The software is easy to use with default parameters. The parameters are tabulated into four groups such as basic, trimming parameters, novel miRNA prediction and other optional parameters.</p></li>
<li><p>Screenshot with basic parameters
<img alt="_images/basic_para.png" src="_images/basic_para.png" /><br/><br/></p></li>
<li><p>Screenshot with trimming parameters
<img alt="_images/trimming_para.png" src="_images/trimming_para.png" /><br/><br/></p></li>
<li><p>Screenshot with novel miRNA predictions
<img alt="_images/novel_para.png" src="_images/novel_para.png" /><br/><br/></p></li>
<li><p>Screenshot with other optional parameters
<img alt="_images/other_para.png" src="_images/other_para.png" /><br/><br/></p></li>
</ul>
</div>
<div class="section" id="resources">
<h2>Resources<a class="headerlink" href="#resources" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Lu, Y., et al., <strong>miRge 2.0 for comprehensive analysis of microRNA sequencing data</strong>. 2018. <em>BMC Bioinformatics</em>. <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/30153801/">PMID</a>.</p></li>
<li><p>Baras, S. A., et al., <strong>miRge - A Multiplexed Method of Processing Small RNA-Seq Data to Determine MicroRNA Entropy</strong>. 2015. <em>PLoS One</em>. <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/26571139/">PMID</a>.</p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="mirge_output.html" class="btn btn-neutral float-right" title="miRge3.0 output" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Arun H. Patil and Marc K. Halushka

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>